### PR TITLE
Binance futures

### PR DESCRIPTION
**Pull request by GuillaumeDSM**
requires https://github.com/Drakkar-Software/OctoBot-Tentacles/pull/996

### CODE REVIEW

File: Trading/Exchange/binance/binance_exchange.py [hunk 1]
```diff
@@ -13,22 +13,57 @@
...
+import octobot_trading.personal_data as trading_personal_data
 
 
 class Binance(exchanges.RestExchange):
     DESCRIPTION = ""
     REQUIRE_ORDER_FEES_FROM_TRADES = True  # set True when get_order is not giving fees on closed orders and fees
     # should be fetched using recent trades.
+    SUPPORTS_SET_MARGIN_TYPE_ON_OPEN_POSITIONS = False  # set False when the exchange refuses to change margin type
+    # when an associated position is open
+    # binance {"code":-4048,"msg":"Margin type cannot be changed if there exists position."}
```
Comment (Herklos) [inline Trading/Exchange/binance/binance_exchange.py:37]: :open_mouth:

File: Trading/Exchange/binance/binance_exchange.py [hunk 2]
```diff
@@ -37,16 +72,166 @@ def get_name(cls):
...
+            params["reduceOnly"] = order.reduce_only
+        return params
+
+    async def create_order(self, order_type: trading_enums.TraderOrderType, symbol: str, quantity: decimal.Decimal,
+                           price: decimal.Decimal = None, stop_price: decimal.Decimal = None,
+                           side: trading_enums.TradeOrderSide = None, current_price: decimal.Decimal = None,
+                           reduce_only: bool = False, params: dict = None) -> typing.Optional[dict]:
+        if self.exchange_manager.is_future:
+            # on futures exchange expects, quantity in contracts: convert quantity into contracts
+            quantity = quantity / self.get_contract_size(symbol)
```
Comment (Herklos) [inline Trading/Exchange/binance/binance_exchange.py:149]: I'm not sure but it seems to be the way to get the quantity for a linear symbol but no the one for an inverse right?

File: Trading/Exchange/bybit/bybit_exchange.py [hunk 3]
```diff
@@ -106,6 +106,27 @@ def get_supported_exchange_types(cls) -> list:
...
+
+    async def _check_unified_account(self):
+        if self.connector.client and not self.exchange_manager.exchange_only:
+            try:
+                self.connector.client.check_required_credentials()
+                enable_unified_margin, enable_unified_account = await self.connector.client.is_unified_enabled()
+                if enable_unified_margin or enable_unified_account:
+                    raise octobot_trading.errors.NotSupported(
+                        "Ignoring Bybit exchange: "
+                        "Bybit unified trading accounts are not yet fully supported. To trade on Bybit, please use a "
```
Comment (Herklos) [inline Trading/Exchange/bybit/bybit_exchange.py:122]: :+1:

**Review by Herklos — COMMENTED** [review]
Great work!

File: Trading/Exchange/binance/binance_exchange.py (hunk 2 above)
Comment (GuillaumeDSM) [reply Trading/Exchange/binance/binance_exchange.py:149]: It's working for both according to my tests:
- linear (contract size on binance is always 1 on linear anyway) BTC/USDT:USDT buy 0.1BTC => contract size = 0.0001, quantity= 0.1 /  0.0001 = 1000 
- inverse BTC/USDT:BTC buy 100 USDT=> contract size = 10, quantity=100 / 10 = 10

Comment (GuillaumeDSM) [reply Trading/Exchange/binance/binance_exchange.py:149]: => and we do the opposite formula when adapting orders from ccxt format to octobot (from contract to settlement asset)

Comment (Herklos) [reply Trading/Exchange/binance/binance_exchange.py:149]: Ok thanks :+1:

File: Trading/Exchange/binance/binance_exchange.py [hunk 4]
```diff
@@ -37,16 +72,166 @@ def get_name(cls):
...
+        else:
+            cls.UNSUPPORTED_ORDERS = cls.SPOT_UNSUPPORTED_ORDERS
+
+    async def get_balance(self, **kwargs: dict):
+        if self.exchange_manager.is_future:
+            balance = []
+            for account_type in self._futures_account_types:
+                balance.append(await self.connector.get_balance(**kwargs, subType=account_type))
+            # todo remove this and use both types when exchange-side multi portfolio is enabled
+            return balance[0]   # only returning linear portfolio
```
Comment (Herklos) [inline Trading/Exchange/binance/binance_exchange.py:134]: Only linear?

Comment (GuillaumeDSM) [reply Trading/Exchange/binance/binance_exchange.py:134]: this comment is wrong, i'll fix it. the idea is to return the 1st fetched (there will only be 1 as long as both linear and inverse are not supported simultaneously)

**Review by Herklos — APPROVED** [review]

**Review by Herklos — APPROVED** [review]

**Review by Herklos — APPROVED** [review]

**Pull request merged**